### PR TITLE
Initial BMFF support

### PR DESCRIPTION
**Pull request by mauricefisher64**
## Changes in this pull request
This change adds initial BMFF file type support.  Writing is limited to MP4 for this PR.  The feature is behind the "bmff" feature flag.

## Checklist
- [x ] This PR represents a single feature, fix, or change.
- [ x] All applicable changes have been documented.
- [ x] Any `TO DO` items (or similar) have been entered as GitHub issues and the link to that issue has been included in a comment.

### CODE REVIEW

File: sdk/src/claim.rs [hunk 1]
```diff
@@ -980,10 +1055,16 @@ impl Claim {
         data_hashes
     }
 
+    pub fn bmff_hash_assertions_mut(&self) -> Vec<&Assertion> {
```
Comment (scouten-adobe) [inline sdk/src/claim.rs:1058]: I don't see anything mutable here. Why the `_mut` suffix?

**Review by scouten-adobe — CHANGES_REQUESTED** [review]
Let's also make sure that we can exclude sdk/tests/fixtures/video1.mp4 file from the crate source when we publish.

crates.io has a hard limit of 10MB per crate that we can not override; I'm reticent to add 3.5MB in one PR.

File: sdk/src/assertions/bmff_hash.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    asset_handlers::bmff_io::bmff_to_jumbf_exclusions,
+    cbor_types::UriT,
+    error::{wrap_io_err, Result},
+    utils::hash_utils::{hash_by_alg, verify_by_alg},
+    Error,
+};
+
+const ASSERTION_CREATION_VERSION: usize = 1;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
```
[the remark this answers is not in the document]
Comment (mauricefisher64) [reply sdk/src/assertions/bmff_hash.rs:37]: This is like DataHash assertions.  It is not user generated.

File: sdk/src/asset_handlers/bmff_io.rs [hunk 3]
```diff
@@ -0,0 +1,1129 @@
...
+
+use atree::{Arena, Token};
+use tempfile::{Builder, NamedTempFile};
+
+use crate::assertions::ExclusionsMap;
+use crate::asset_io::{AssetIO, CAILoader, CAIRead, HashObjectPositions};
+use crate::error::{Error, Result};
+use crate::utils::hash_utils::{vec_compare, Exclusion};
+
+pub struct BmffIO {
```
[the remark this answers is not in the document]
Comment (mauricefisher64) [reply sdk/src/asset_handlers/bmff_io.rs:33]: Who does :)

File: sdk/src/claim.rs (hunk 1 above)
Comment (mauricefisher64) [reply sdk/src/claim.rs:1058]: Looks like I changed the usage.  I will fix the name.

**Review by gpeacock — APPROVED** [review]

File: sdk/src/claim.rs [hunk 4]
```diff
@@ -916,10 +956,45 @@ impl Claim {
             }
 
             for dh_assertion in claim.data_hash_assertions() {
-                let dh = DataHash::from_assertion(&dh_assertion)?;
-                let name = dh.name.as_ref().map_or("unnamed".to_string(), default_str);
-                if !dh.is_remote_hash() {
-                    // only verify local hashes here
+                if dh_assertion.label_root() == DataHash::LABEL {
+                    let dh = DataHash::from_assertion(dh_assertion)?;
+                    let name = dh.name.as_ref().map_or("unnamed".to_string(), default_str);
```
[the remark this answers is not in the document]
Comment (mauricefisher64) [reply sdk/src/claim.rs]: done

File: sdk/src/jumbf_io.rs [hunk 5]
```diff
@@ -12,26 +12,59 @@
...
 use std::{
-    fs,
+    fs::{self, File},
     io::Cursor,
     path::{Path, PathBuf},
 };
 
-static SUPPORTED_TYPES: &[&str; 6] = &[
+static SUPPORTED_TYPES: [&str; 18] = [
+    "avif",
```
Comment (scouten-adobe) [inline sdk/src/jumbf_io.rs:27]: I think we have a design issue here: Some of these file formats are only available if you've enabled the `bmff` crate feature, right? Not a blocker for this PR, but I'd like to rethink how we describe what is and isn't supported in the face of multiple feature sets.

**Review by scouten-adobe — APPROVED** [review]

File: sdk/src/jumbf_io.rs (hunk 5 above)
Comment (mauricefisher64) [reply sdk/src/jumbf_io.rs:27]: @scouten-adobe   We don't currently have a way to advertise supported formats.  Also some may be for reading, writing,  in-memory, wasm, or other.

File: sdk/src/asset_handlers/bmff_io.rs [hunk 6]
```diff
@@ -1064,9 +1064,10 @@ impl AssetIO for BmffIO {
 
 #[cfg(test)]
 #[cfg(feature = "bmff")]
-#[allow(clippy::expect_used)]
```
Comment (mauricefisher64) [inline sdk/src/asset_handlers/bmff_io.rs]: @scouten-adobe Is this needed?  Seems to have broken the build.

**Pull request merged**